lib/netext/resolver.go

```diff
@@ -162,21 +149,26 @@ func (r *resolver) selectOne(host string, ips []net.IP) net.IP {
 	return ip
 }
 
-func (r *resolver) applyPolicy(host string, ip4, ip6 []net.IP) (ip net.IP, ips []net.IP) {
-	switch r.policy {
+func (r *resolver) applyPolicy(ips []net.IP) (retIPs []net.IP) {
+	if r.policy == lib.DNSany {
+		return ips
+	}
+	ip4, ip6 := groupByVersion(ips)
+	switch r.policy { //nolint: exhaustive
 	case lib.DNSpreferIPv4:
-		ip = r.selectOne(host, ip4)
-		ips = ip6
+		retIPs = ip6
+		if len(ip4) > 0 {
+			retIPs = ip4
+		}
```

Comment:
nit: I find it strange that preferIPv4 first doesn't  set retIPs to ipv4 and then if retIPs is empty to ... actually set it to ipv6

Reply:
Sure, fixed in the latest push.